corehq/apps/users/login_record.py

```diff
@@ -0,0 +1,83 @@
+from datetime import datetime
+from django_redis import get_redis_connection
+
+ONE_DAY = 60 * 60 * 24
+EPOCH = datetime.utcfromtimestamp(0)
+
+
+class LoginRecord:
+    """A thread-safe class to record login failures."""
+
+    FIELD_FAILURES = 'failures'
+    FIELD_TS = 'ts'
+
+    @classmethod
+    def get(cls, username):
+        raw_failures, raw_ts = cls._get_raw_fields(username)
+        attempts_ts = float(raw_ts or 0)
+        num_failures = int(raw_failures or 0)
+
+        last_attempt_date = datetime.utcfromtimestamp(attempts_ts or 0)
+
+        return LoginRecord(username, num_failures, last_attempt_date)
+
+    @classmethod
+    def _get_raw_fields(cls, username):
+        redis_client = get_redis_connection()
+        return redis_client.hmget(cls._get_key(username), cls.FIELD_FAILURES, cls.FIELD_TS)
+
+    @staticmethod
+    def _get_key(username):
+        return f'attempts_{username}'
+
+    def __init__(self, username, failures=0, last_attempt_date=None):
+        self.username = username
+        self._reset_values(failures, last_attempt_date)
+
+        self.key = self._get_key(self.username)
+
+    def _reset_values(self, failures=0, last_attempt_date=None):
+        self.failures = failures
+        self.last_attempt_date = last_attempt_date or EPOCH
+
+    def clear(self):
+        redis_client = get_redis_connection()
+        redis_client.delete(self.key)
+
+        self._reset_values()
+
+    def add_failure(self, current_time):
+        redis_client = get_redis_connection()
+        transaction_results = redis_client.transaction(
+            self._create_failure_handler(current_time),
+            self.key)
```

Comment:
Have to say, I'm learning a lot about redis reading this PR. I've mostly used just the most basic operations in redis like get, set, incr, etc. So this is neat.

According to https://redis-py.readthedocs.io/en/stable/index.html?highlight=transaction#redis.Redis.transaction `.transaction` is a way to call `WATCH ... MULTI ... EXEC ...` in redis (also just learned about this here https://redis.io/topics/transactions). Watch makes the exec fail and return none if they watch key has been updated by another query. I can't tell what `.transaction` returns though, and whether you're checking for that here or not. The docs I linked suggest that with WATCH the common pattern is to keep trying it in a loop until it goes through. Is the idea that .transaction does that loop for you? Kind of looks like it in the source code linked https://redis-py.readthedocs.io/en/stable/_modules/redis/client.html#Redis.transaction.

Reply:
Yep! I agree that the convenience method doesn't make it particularly obvious what is going on, but transaction is just using optimistic locking, looping until `WATCH` doesn't have a mismatch, at which point it returns the results.